src/mainwindow/mainwindow.cpp

```diff
@@ -427,18 +601,92 @@ void MainWindow::onCheckUpdatesTriggered()
 
 void MainWindow::onOpenFileFromProjectViewer(QString fileName)
 {
+    // if document already opened then return
+    if(checkIfOpened(fileName))
+    {
+        QMessageBox::warning(this, "Document already opened", "Selected document already opened.");
+        return;
+    }
+
     openDoc(fileName);
 }
 
+void MainWindow::onCloseWindow(CodeEditor *curDoc)
+{
+    if(!curDoc)
+        return;
+
+    if(curDoc->document()->isModified())
+    {
+        QMessageBox::StandardButton reply = QMessageBox::question
+                (
+                    this,
+                    "Saving Changes",
+                    "Do you want to save changes to opened documents?",
+                    QMessageBox::Yes | QMessageBox::No
+                    );
+
+        if(reply == QMessageBox::No)
+            return;
+
+        saveDocument(curDoc, curDoc->getFileName());
+    }
+}
+
 CodeEditor* MainWindow::createNewDoc()
 {
     CodeEditor *newDoc = new CodeEditor;
```

Comment:
I suppose you will have memory leaks. Generally, it is bad practice to create an object by new and return an unmanaged pointer. If someone uses your function, he won't know about new and won't use delete.

Reply:
We tried to follow Qt memory management approach
Every instance of CodeEditor is passed to parent class MDIArea.
Qt guarantees that memory will be freed.